x-pack/plugins/fleet/server/routes/settings/enrollment_settings_handler.ts

```diff
@@ -0,0 +1,185 @@
+import { getFleetServerHostsForAgentPolicy } from '../../services/fleet_server_host';
+import { getFleetProxy } from '../../services/fleet_proxies';
+
+export const getEnrollmentSettingsHandler: FleetRequestHandler<
+  undefined,
+  TypeOf<typeof GetEnrollmentSettingsRequestSchema.query>
+> = async (context, request, response) => {
+  const agentPolicyId = request.query?.agentPolicyId;
+  const settingsResponse: GetEnrollmentSettingsResponse = {
+    fleet_server: {
+      agent_policies: [],
+      has_active: false,
+    },
+  };
+  const coreContext = await context.core;
+  const esClient = coreContext.elasticsearch.client.asInternalUser;
+  const soClient = coreContext.savedObjects.client;
+  try {
+    // Get all possible fleet server or scoped normal agent policies
+    const { fleetServerAgentPolicies, scopedAgentPolicy: scopedAgentPolicyResponse } =
+      await getFleetServerAgentPolicies(soClient, agentPolicyId);
+    const scopedAgentPolicy = scopedAgentPolicyResponse || {
+      id: undefined,
+      name: undefined,
+      fleet_server_host_id: undefined,
+      download_source_id: undefined,
+    };
+
+    // Check if there is any active fleet server enrolled into the fleet server policies policies
+    if (fleetServerAgentPolicies) {
+      settingsResponse.fleet_server.agent_policies = fleetServerAgentPolicies;
+      settingsResponse.fleet_server.has_active = await hasActiveFleetServersForAgentPolicies(
+        esClient,
+        soClient,
+        fleetServerAgentPolicies.map((p) => p.id)
+      );
+    }
+
+    // Get download source
+    // ignore errors if the download source is not found
+    try {
+      settingsResponse.download_source = await getDownloadSource(
+        soClient,
+        scopedAgentPolicy.download_source_id ?? undefined
+      );
+    } catch (e) {
+      settingsResponse.download_source = undefined;
+    }
+
+    // Get associated fleet server host, or default one if it doesn't exist
+    // `getFleetServerHostsForAgentPolicy` errors if there is no default, so catch it
+    try {
+      settingsResponse.fleet_server.host = await getFleetServerHostsForAgentPolicy(
+        soClient,
+        scopedAgentPolicy
+      );
+    } catch (e) {
+      settingsResponse.fleet_server.host = undefined;
+    }
+
+    // if a fleet server host was found, get associated fleet server host proxy if any
+    // ignore errors if the proxy is not found
+    try {
+      if (settingsResponse.fleet_server.host?.proxy_id) {
+        settingsResponse.fleet_server.host_proxy = await getFleetProxy(
+          soClient,
+          settingsResponse.fleet_server.host.proxy_id
+        );
+      }
+    } catch (e) {
+      settingsResponse.fleet_server.host_proxy = undefined;
+    }
+
+    return response.ok({ body: settingsResponse });
+  } catch (error) {
+    return defaultFleetErrorHandler({ error, response });
+  }
+};
+
+const getFleetServerAgentPolicies = async (
```

Comment:
I know that we have integration tests, but would it be possible to add some unit tests for these utility functions? They do a lot of things and it's easy to miss some edge case.

Reply:
good idea, I will add

Reply:
added!